### DISCUSSION TITLE
Firebase Crashlytics on iOS devices cannot communicate (not integrated)

### DISCUSSION THREAD

**Discussion by firdiansyah26** [invertase/react-native-firebase#4227] [Q&A] [answered]
Hello everyone, I'm having a problem installing the **@react-native-firebase/crashlytics** module on my iOS device. I have added **GoogleService-Info.plist (Figure 1)** along with the configuration in the **AppDelegate.m (Figure 2)** and **Podfile (Figure 3)** files. But on the Firebase Crashlytics web there is still loading continuously without any error messages **(figure 4)**, and firebase.json **(figure 5)**, any solution? _(I'm using objective-c)_

Thank You

**_Detail_**
Figure 1 _(GoogleService-Info.plist)_
![image](https://user-images.githubusercontent.com/36752128/92428039-5163e880-f1b8-11ea-8f2a-b2ec4b27d036.png)
![image](https://user-images.githubusercontent.com/36752128/92428805-588bf600-f1ba-11ea-87b8-fa6d4e083ff4.png)
![image](https://user-images.githubusercontent.com/36752128/92428886-8ffaa280-f1ba-11ea-97cc-c42e102cec7d.png)

Figure 2 _(AppDelegate.m)_
![image](https://user-images.githubusercontent.com/36752128/92428095-735d6b00-f1b8-11ea-9ef5-084309286792.png)
![image](https://user-images.githubusercontent.com/36752128/92428160-8e2fdf80-f1b8-11ea-89a6-1be8ea55fed6.png)
![image](https://user-images.githubusercontent.com/36752128/92428243-c505f580-f1b8-11ea-9d08-f099c1eb2932.png)

Figure 3 _(Podfile)_
![image](https://user-images.githubusercontent.com/36752128/92428300-ea92ff00-f1b8-11ea-85c2-9be01b6f8af1.png)

Figure 4 _(Firebase Crashlytics UI web)_
![image](https://user-images.githubusercontent.com/36752128/92428644-dc91ae00-f1b9-11ea-8d0d-8ed23bd1bd7b.png)

Figure 5 _(Firebase.json)_
![image](https://user-images.githubusercontent.com/36752128/92450475-aff38b80-f1e5-11ea-999f-4ca522140343.png)

**Comment by mikehardy** [comment]
This is very frustrating yes! If I remember correctly this happens when you are doing crashlytics without analytics. I might be wrong, but I think if you do create a test screen with a button that calls the crash function, then restart the app (that's when it actually sends it! the next time it runs is the first chance it gets...) then wait a bit, it should show up. If it is a debug build you will also need to to make sure your firebase.json has the right stuff: https://github.com/invertase/react-native-firebase/issues/3879#issuecomment-678389795

If you do all that...you should get past this screen and see a crash in the console. If nothing else works you might temporarily try installing the analytics package as well

I am sorry I can't locate the upstream issue where this is discussed so I do not have a great pointer on it with hard facts, but if I am right that it is just an interaction with crashlytics and analytics hopefully a test crash gets you past it so the above helps

**Reply by firdiansyah26** [reply]
I've implemented Firebase Analytics but it's still the same ._.

**Comment by mikehardy** [comment]
@firdiansyah26 did you do a test crash and make sure it would upload?

**Reply by firdiansyah26** [reply]
@mikehardy yes, I have used crash tests using `crashlytics().crash()` from `import crashlytics from '@react-native-firebase/crashlytics';`, but still nothing has changed

**Comment by williamhaley** [comment]
First time Crashlytics user with RNFirebase having the same issue. I can't get any crash reports/errors/exceptions to show up in Firebase. I've been trying for several days with no luck. I generated an Ad-Hoc `.ipa` archive with the `Release` configuration and am purposefully triggering crashes with...

```
crashlytics().recordError(new Error('test'));
crashlytics().crash();
```

...but nothing ever appears in the Crashlytics console in Firebase. I had followed the [rnfirebase app](https://rnfirebase.io/) and [rnfirebase crashlytics](https://rnfirebase.io/crashlytics/usage) setup guides, tried using the simulator + a physical device with the Release config, explicitly killed the app to make sure it's not attached to the debugger [[source]](https://stackoverflow.com/questions/54464214/react-native-firebase-crashlytics-not-showing-up-on-firebase-dashboard), setting `DWARF with dSYM File` [[source]](https://firebase.google.com/docs/crashlytics/test-implementation?platform=ios), nothing seems to work in all my troubleshooting. I can't see anything in the Crashlytics dashboard aside from the infinite, "Add the SDK, then build, run, and crash your app", loading indicator. The console never advances beyond that.

Even with the release build I have the following `firebase.json` in case that knocks something loose, but no joy.

```
{
  "react-native": {
    "crashlytics_debug_enabled": true
  }
}
```

The library seems to be installed properly. The AppDelegate `[FIRApp configure];` seems to work, and I'm able to use the Crashlytics API in React Native without any runtime issues. I feel like I'm missing something basic, but it's unfortunately been driving me bonkers in the mean time.

**Reply by firdiansyah26** [reply]
Hi @williamhaley, yes, I feel it, easy to implement on Android but not on iOS, I've tried uninstalling the bundler package in the firebase console, and restarting the installation following the existing document, but nothing happened, the console crashlytics still shows loading. Is your iOS app running Firebase Crashlytics successfully now?

**Reply by williamhaley** [reply]
Unfortunately not @firdiansyah26. I'm still in the midst of trying to get it to work. The next thing I plan to try is simply working with a [generic iOS Crashlytics setup](https://firebase.google.com/docs/crashlytics/get-started?platform=ios) and triggering a crash directly from Objective-C. My hope/thinking is that I can 1) verify I can get *anything* to work with Crashlytics in a more vanilla form and 2) maybe if that succeeds, subsequent RNFirebase errors will start getting logged.

**Reply by firdiansyah26** [reply]
Thank you for sharing @williamhaley, I also plan to ask the iOS developer this, hoping to get input on this issue, good luck!.

**Comment by mikehardy** [comment]
I'm sorry I don't have a solid solution on this one, but I'm definitely listening - I have seen what you describe before I just don't recall anyone posting solid steps to get past it. Very interested in the experience after a vanilla iOS implementation - my hunch is there is either something basic we're missing (like perhaps a temporary analytics integration of some sort is necessary but you have to do something specific within the two packages) or that upstream really has an issue. Either way, I'm listening intently and we can document whatever findings result

**Reply by williamhaley** [reply]
No need to apologize @mikehardy! Your hard work is appreciated.

For debugging, `Product` -> `Scheme` -> `Edit Scheme...` -> `Run` -> `Arguments` -> `Arguments Passed On Launch` -> `+` and adding `-FIRDebugEnabled` was extremely helpful! [[source]](https://stackoverflow.com/questions/43109486/turning-firebase-analytics-on-on-xcode)

```
2020-09-09 12:52:58.649940-0500 Changes[32761:5065526] 6.30.0 - [Firebase/Crashlytics][I-CLS000000] Automatic data collection is disabled.
2020-09-09 12:52:58.669140-0500 Changes[32761:5065515] 6.30.0 - [Firebase/Crashlytics][I-CLS000000] [Crashlytics:Crash] 40 unsent reports are available. Checking for upload permission.
2020-09-09 12:52:58.669899-0500 Changes[32761:5065515] 6.30.0 - [Firebase/Crashlytics][I-CLS000000] [Crashlytics:Crash] Notifying that unsent reports are available.
2020-09-09 12:52:58.670492-0500 Changes[32761:5065515] 6.30.0 - [Firebase/Crashlytics][I-CLS000000] [Crashlytics:Crash] Waiting for send/deleteUnsentReports to be called.
```

The `40 unsent reports are available. Checking for upload permission.` caught my eye first. I didn't even notice the `Automatic data collection is disabled.` initially. It was clear to me that my install was collecting crashes but it seemed to never be uploading them.

A search for that led me to a [StackOverflow Question](https://stackoverflow.com/questions/62788245/how-to-force-crashlytics-to-upload-reports-wait), which led me to some [Firebase docs that were updated very recently](https://firebase.google.com/docs/crashlytics/upgrade-sdk?platform=ios#the_crashlyticsdelegate_is_replaced_by_separate_methods_for_handling_crash_reports).

This seems most relevant.

```
[[FIRCrashlytics crashlytics] setCrashlyticsCollectionEnabled:false];

[[FIRCrashlytics crashlytics] checkForUnsentReportsWithCompletion:^(BOOL hasUnsentReports) {
  BOOL hasConsent = false;
  // ...get consent from user.

  if (hasConsent && hasUnsentReports) {
    [[FIRCrashlytics crashlytics] sendUnsentReports];
  } else {
    [[FIRCrashlytics crashlytics] deleteUnsentReports];
  }
}];

// Detect when a crash happens during your app's last run.
if ([[FIRCrashlytics crashlytics] didCrashDuringPreviousExecution]) {
  // ...notify the user.
}
```

If I logged `isCrashlyticsCollectionEnabled` it was coming back as `0` (false). I had to explicitly enable it. Note that there seem to be new patterns around enabling explicit data collection for crashlytics and hooks allowing app developers to let users consent to crash reports being uploaded.

I simply changed `[[FIRCrashlytics crashlytics] setCrashlyticsCollectionEnabled:false];` to `[[FIRCrashlytics crashlytics] setCrashlyticsCollectionEnabled:true];` in that sample code and pasted it into my `AppDelegate.m` and voila, my crash reports all uploaded! This is what the top of my `AppDelegate.m` looks like now.

```
- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions
{
  [FIRApp configure];
  
  NSLog(@"Is collection enabled? %d", [FIRCrashlytics crashlytics].isCrashlyticsCollectionEnabled);
  
  [[FIRCrashlytics crashlytics] setCrashlyticsCollectionEnabled:true];
```

Not sure if this is a very recent change that caught me and @firdiansyah26 or if this is somehow specific to my install.

Either way, things are **working now!** By calling `setCrashlyticsCollectionEnabled` my crash reports are finally showing up online. No other changes needed on my end. This worked while running on-device with the debugger attached.

**Edit**: In hindsight, not sure if these docs I reference are actually "new" or "recent". They fall under the heading `Legacy Fabric SDK Guides` and say `If you recently migrated your app from Fabric...`, which I don't think applies to me. I guess the most significant takeaway was that crash collection was *not* enabled for me, and enabling it fixed things. No idea why it was needed.

**Reply by williamhaley** [reply]
![Screen Shot 2020-09-09 at 1 54 53 PM](https://user-images.githubusercontent.com/1851748/92641242-117a2f80-f2a4-11ea-853d-fa9eb8c7fc56.png)

**Reply by firdiansyah26** [reply]
Hi @williamhaley, I have tried what you say, and got a console message like this, but for web firebase crashlytics is still like the beginning :(

```
9:26.217137+0700 [--YourApp--][5766:2556448] [Firebase/Crashlytics] Version 4.4.0
2020-09-10 10:19:26.218015+0700 [--YourApp--][5766:2556695] 6.30.0 - [Firebase/Crashlytics][I-CLS000000] [Crashlytics] Running on iPhone12,1, 13.6.1 (17G80)
2020-09-10 10:19:26.218525+0700 [--YourApp--][5766:2556695] 6.30.0 - [Firebase/Crashlytics][I-CLS000000] Root: /var/mobile/Containers/Data/Application/778E5A5E-879D-42C4-9F4B-D069F614E044/Library/Caches/com.crashlytics.data/[--BundleId--]
2020-09-10 10:19:26.228158+0700 [--YourApp--][5766:2556715] 6.30.0 - [Firebase/Crashlytics][I-CLS000000] [Crashlytics:Crash:Reports:Event] Registered Firebase Analytics event listener
2020-09-10 10:19:26.228434+0700 [--YourApp--][5766:2556715] 6.30.0 - [Firebase/Crashlytics][I-CLS000000] Automatic data collection is enabled.
2020-09-10 10:19:26.228595+0700 [--YourApp--][5766:2556715] 6.30.0 - [Firebase/Crashlytics][I-CLS000000] Unsent reports will be uploaded at startup
2020-09-10 10:19:26.276379+0700 [--YourApp--][5766:2556721] 6.30.0 - [Firebase/Crashlytics][I-CLS000000] Image not found
2020-09-10 10:19:26.277266+0700 [--YourApp--][5766:2556721] 6.30.0 - [Firebase/Crashlytics][I-CLS000000] Image not found
2020-09-10 10:19:26.287357+0700 [--YourApp--][5766:2556716] 6.30.0 - [Firebase/Crashlytics][I-CLS000000] Image not found
2020-09-10 10:19:26.287400+0700 [--YourApp--][5766:2556716] 6.30.0 - [Firebase/Crashlytics][I-CLS000000] Image not found
2020-09-10 10:19:26.287469+0700 [--YourApp--][5766:2556716] 6.30.0 - [Firebase/Crashlytics][I-CLS000000] Image not found
2020-09-10 10:19:26.287514+0700 [--YourApp--][5766:2556716] 6.30.0 - [Firebase/Crashlytics][I-CLS000000] Image not found
2020-09-10 10:19:26.387695+0700 [--YourApp--][5766:2556696] [Firebase/Crashlytics] Packaged report with id '8ca10e3bf3a04452b3cb852090f1f9c9' for submission
2020-09-10 10:19:26.389642+0700 [--YourApp--][5766:2556696] 6.30.0 - [Firebase/Crashlytics][I-CLS000000] Preparing the report for the new endpoint: 1
2020-09-10 10:19:26.389680+0700 [--YourApp--][5766:2556696] 6.30.0 - [Firebase/Crashlytics][I-CLS000000] [Crashlytics:Crash:Reports] Submitting report
:26.389708+0700 [--YourApp--][5766:2556696] 6.30.0 - [Firebase/Crashlytics][I-CLS000000] [Crashlytics:Crash:Reports:Event] Sending event.
2020-09-10 10:19:26.392054+0700 [--YourApp--][5766:2556716] 6.30.0 - [Firebase/Crashlytics][I-CLS000000] [Crashlytics:Crash:Reports] Completed report submission with id: 8ca10e3bf3a04452b3cb852090f1f9c9
```

**Reply by williamhaley** [reply]
Hmm 🤔 Can you try setting `  [[FIRCrashlytics crashlytics] setCrashlyticsCollectionEnabled:false];` (false) and triggering a bunch of crashes? Eventually you should see a number of unsent crash reports.

```
40 unsent reports are available. Checking for upload permission.
```

That would at least indicate that Crashlytics is properly seeing and collecting crashes locally, even if it's not transmitting them.

I sometimes found it was tricky to trigger a crash appropriately in Debug mode. I had to make sure the debugger was not attached (stop the app in Xcode then launch the app manually).

**Reply by mikehardy** [reply]
@williamhaley has good advice here to just verify for sure that logging is happening but I think this may be a separate problem related to projects that were on Fabric originally (if I'm following the thread - if that issue is already resolved, sorry)

**Reply by firdiansyah26** [reply]
still same, maybe you can correction my code _*cmiiw_

1. AppDelegate.m
![image](https://user-images.githubusercontent.com/36752128/92682010-6b88fc80-f359-11ea-9596-a12c369cece3.png)

2. Podfile
![image](https://user-images.githubusercontent.com/36752128/92682029-7c397280-f359-11ea-8702-077186314def.png)

3. Build Phase
![image](https://user-images.githubusercontent.com/36752128/93447581-c217ad00-f8fc-11ea-98a0-6f12969e7316.png)

4. Scheme
![image](https://user-images.githubusercontent.com/36752128/93447704-e8d5e380-f8fc-11ea-9037-622313423bf4.png)

![image](https://user-images.githubusercontent.com/36752128/93447923-318d9c80-f8fd-11ea-8bc9-07b88f439e5b.png)

5. Test Crash Function
![image](https://user-images.githubusercontent.com/36752128/92682217-0255b900-f35a-11ea-848b-7e15b5a618ad.png)
 
6. Firebase Json
![image](https://user-images.githubusercontent.com/36752128/92718948-b40ddd80-f38c-11ea-9ca0-eb7e86652fbc.png)

**Reply by williamhaley** [reply]
@firdiansyah26 Can you try purposefully crashing your app in `AppDelegate.m`? Make sure the debugger's detached (press stop) then launch your app manually (just by tapping the icon on your phone) to run and crash it. You can throw this error in the bottom of `didFinishLaunchingWithOptions` in your `AppDelegate.m`.

```
// This is the same type of exception that gets thrown in JS by calls to crashlytics().crash();
@throw NSInternalInconsistencyException;
```

Also, I find that these log messages are most helpful. Do you ever see these?

```
2020-09-13 22:49:32.101200-0500 THEAPP[4057:573254] 6.30.0 - [Firebase/Crashlytics][I-CLS000000] Automatic data collection is enabled.
2020-09-13 22:49:32.101694-0500 THEAPP[4057:573254] 6.30.0 - [Firebase/Crashlytics][I-CLS000000] Unsent reports will be uploaded at startup
```

I have a tough time re-creating reliable crashes on my end. It seems like exceptions I throw from Objective-C or Swift show up quickly so at least those are mostly reliable (even though the stack traces seem to take hours to appear at times), but (for reasons unclear or due to lack of understaning on my part) exceptions thrown from JS/React code either never show up or take a long time to populate in the dashboard.

From what I can tell it's no fault of react-native-firebase. Rather, Crashlytics seems to take forever to sometimes show my errors, and their UI seems to be inconsistent as I debug which complicates things. I have to admit, although it's nice to have everything in one spot with Firebase + Crashlytics, I'm probably going to move to Bugsnag. I've used them in the past and they have a few features I'm missing. Good luck!

**Reply by firdiansyah26** [reply]
Hi @williamhaley, yes I saw a message in the console, I also repeated the steps in the documentation and applied it to a new project, but it's the same, loading is still running which indicates the integration has not been successful. Maybe for iOS it will be like now and I will ask another discussion forum regarding this matter, hopefully there are suggestions for this problem. Thank you for your help sir!

**Comment by mikehardy** [comment]
These are very recent APIs, note that we support them as of recently as well: https://github.com/invertase/react-native-firebase/pull/4009

But it's supposed to be enabled with this: `{ "react-native": { "crashlytics_disable_auto_disabler": true, "crashlytics_debug_enabled": true } }` from the linked comment above

Perhaps things moved yet again with regard to defaults

It is still a pending work item for me to have crashlytics turned *on* by default in debug and release so it works out of the box, and then document how to disable it, as opposed to this current state where you have to turn on like 3 toggles to get debug crashes to upload.

That has not gotten done in weeks though, so if you're still in there and you know where the settings are in our code, if you could do a PR to have iOS log crashes by default, in debug too, that would clearly help a lot of people :pray:

**Reply by williamhaley** [reply]
Will do! I've been digging into the source and learning a lot about RNFirebase. I'll post my notes when I have something conclusive or am ready to get a PR together.

**Reply by williamhaley** [reply]
See https://github.com/invertase/react-native-firebase/pull/4236 where I _think_ I did the appropriate work to get iOS crash collection enabled by default. I actually hadn't previously seen `crashlytics_disable_auto_disabler` until you mentioned it here. I wasn't sure what I was looking for in #4009. Did you mean https://github.com/invertase/react-native-firebase/issues/3879? I only just found that now (whoops!) and see that a lot of people have already trod on the ground I walked today. Darn, hopefully the work I did is worthwhile based on previous design conversations.

**Comment by firdiansyah26** [comment, marked as answer]
**Case is closed**

The configuration in my code is like this

1. package.json
 ` "@react-native-firebase/crashlytics": "^8.4.1"` => `"@react-native-firebase/crashlytics": "8.3.0"`

2. Podfile
   `$FirebaseSDKVersion = '6.30.0'`

3. Info.plist
   `<key>FirebaseCrashlyticsCollectionEnabled</key>` set => `<true/>`

![image](https://user-images.githubusercontent.com/36752128/93297857-55c57c80-f81c-11ea-9a94-19952c715877.png)


Thank you for your contribute for this case, @williamhaley @mikehardy
